### PR TITLE
Create consignment browse view with tests and introduce factory boy

### PR DESCRIPTION
**Pull request by anthonyhashemi**
## Changes in this PR

- Made consignment browse page work, inlcuding 
      - Added consignment browse view query
       - Hooked up to browse route
        - Adjusted browse template for when consignment_id passed

- Added factory boy to repo, including
      - Created Factory Classes for all of our SQLAlchemy Model classes.

- Added tests for all of the new logic and functionality

## JIRA ticket

https://national-archives.atlassian.net/browse/AYR-526

## Screenshot of UI

Screenshot 2023-12-21 at 12 29 41

### CODE REVIEW

**Review by pgandhizaizi — APPROVED** [review]
looks good, I can see file type filter has been added, no more changes required, approved

**Comment by anthonyhashemi** [comment]
HOLDING OFF TO MERGE ONLY ONCE the https://github.com/nationalarchives/da-ayr-beta-webapp/pull/99 is merged in

**Comment by pgandhizaizi** [comment]
was wondering that you current consignment ref query will support transferring body and series filters ?

File: app/tests/test_browse.py [hunk 1]
```diff
@@ -357,19 +357,28 @@ def test_browse_consignment(client: FlaskClient):
...
-    # ]
-
-    # assert [header.text.replace('\n', ' ').strip(' ') for header in headers] == expected_results_table[0]
-    # row_data = ''
-    # for row_index, row in enumerate(rows):
-    #    row_data = row_data + "'" + row.text.replace('\n', ' ').strip(' ') + "'"
-    #    if row_index < len(rows) - 1:
-    #        row_data = row_data + ", "
-    # assert [row_data] == expected_results_table[1]
+    soup = BeautifulSoup(response.data, "html.parser")
```
Comment (anthonyhashemi) [inline app/tests/test_browse.py:360]: test for the browse consignment route itself

File: app/tests/test_queries.py [hunk 2]
```diff
@@ -392,6 +397,91 @@ def test_browse_data_with_series_filter(self, client: FlaskClient):
             )
         ]
 
+    def test_browse_data_with_consignment_filter(self, app):
```
Comment (anthonyhashemi) [inline app/tests/test_queries.py:400]: test for the browse consignment query, to show it filters objects as expected

File: app/tests/test_template_filters.py [hunk 3]
```diff
@@ -0,0 +1,12 @@
+from app.template_filters import format_datetime
+
+
+class TestFormatDateTime:
```
Comment (anthonyhashemi) [inline app/tests/test_template_filters.py:4]: tests for the new template filter `format_datetime`

File: app/tests/test_queries.py [hunk 4]
```diff
@@ -392,6 +397,91 @@ def test_browse_data_with_series_filter(self, client: FlaskClient):
...
+            where 2 file types is 'file', another file 'folder', and another file of
+            type 'file' and metadata associated with a different consignment
+        When I call the 'browse_data' function with the consignment id
+            of the first 3 file objects
+        Then it returns a Pagination object with 2 total results corresponding to the
+            first 2 files, ordered by their names and with the expected metadata values
+        """
+        consignment = ConsignmentFactory()
+
+        file_1 = FileFactory(
```
Comment (anthonyhashemi) [inline app/tests/test_queries.py:412]: utilising factory boy which I introduced as a dev dependency. Much cleaner test setup.

File: app/tests/factories.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
+import uuid
+
+import factory
+from factory.alchemy import SQLAlchemyModelFactory
+from factory.fuzzy import FuzzyText
+
+from app.main.db.models import Body, Consignment, File, FileMetadata, Series, db
+
+
+class BodyFactory(SQLAlchemyModelFactory):
```
Comment (anthonyhashemi) [inline app/tests/factories.py:10]: this file defines all of our factory boy SQLAlchemyModelFactories, each corresponding to our SQLAlchemy Model classes defining our db tables.

File: app/templates/main/browse.html [hunk 6]
```diff
@@ -335,59 +335,56 @@ <h1 class="govuk-heading-l browse__records-found__text">Records found {{ num_rec
...
         {% endif %}
       {% endif %}
 
-      {% if num_records_found > 0 %}
-        <!-- PAGINATION -->
-        {% with view_name='main.browse' %}
-          {% include "pagination.html" %}
-        {% endwith %}
-      {% endif %}
+      <!-- PAGINATION -->
```
Comment (anthonyhashemi) [inline app/templates/main/browse.html:384]: pagination tremplate has condition of total pages > 1 inside it anyway so this condition was redundant

File: app/templates/main/browse.html [hunk 7]
```diff
@@ -335,59 +335,56 @@ <h1 class="govuk-heading-l browse__records-found__text">Records found {{ num_rec
           {% if num_records_found > 0 %}
             <dl class="govuk-summary-list browse-grid__list">
               <div class="govuk-summary-list__row govuk-summary-list__row__browse-grid-list">
-                <dt class="govuk-summary-list__key browse-grid__key browse-grid__key__transferring-body">
+                <dt class="govuk-summary-list__key govuk-summary-list__key__left">
```
Comment (anthonyhashemi) [inline app/templates/main/browse.html:338]: had to do this to make the columns align properly. Not quite sure why we wanted to right align them previously anyway

File: app/template_filters.py [hunk 8]
```diff
@@ -0,0 +1,23 @@
+import datetime
+
+
+def format_datetime(value: str) -> str:
```
Comment (anthonyhashemi) [inline app/template_filters.py:4]: Use this in the browse.html template to format datestrings as formatted in the db to how we want them to be presnted in html

File: app/static/src/scss/record.scss [hunk 9]
```diff
@@ -139,6 +139,10 @@
         width:0;
         text-align:right;
         font-size: 1rem;
+
+        &__left {
```
Comment (anthonyhashemi) [inline app/static/src/scss/record.scss:143]: needed to left align for my table - didnt want to break all the others if they need to be right aligned for some reason

File: app/main/routes.py [hunk 10]
```diff
@@ -99,19 +99,23 @@ def browse():
...
-            page, per_page, transferring_body_id=transferring_body_id
-        )
+        filters = {"transferring_body_id": transferring_body_id}
     elif series_id:
         browse_type = "series"
-        browse_results = browse_data(page, per_page, series_id=series_id)
+        filters = {"series_id": series_id}
+    elif consignment_id:
+        browse_type = "consignment"
+        filters = {"consignment_id": consignment_id}
```
Comment (anthonyhashemi) [inline app/main/routes.py:114]: refactored this whilst adding the new consignent_id filter

File: app/main/db/queries.py [hunk 11]
```diff
@@ -237,3 +247,60 @@ def _build_series_filter_query(series_id):
     )
 
     return query
+
+
+def _build_consignment_filter_query(consignment_id: uuid.UUID):
```
Comment (anthonyhashemi) [inline app/main/db/queries.py:252]: consignment browse query, retrieving only the fields we want to present in the FE as well as the File ID

File: app/__init__.py [hunk 12]
```diff
@@ -100,6 +101,8 @@ def create_app(config_class, database_uri=None):
 
     app.register_blueprint(main_bp)
 
+    app.jinja_env.filters["format_datetime"] = format_datetime
```
Comment (anthonyhashemi) [inline app/__init__.py:104]: register the template filter on the app

**Comment by anthonyhashemi** [comment]
> was wondering that you current consignment ref query will support transferring body and series filters ?

The query can easly be extended to filter by those later if we need. When we come to doing the filter ticket(s) that can be adjusted

**Review by pgandhizaizi — APPROVED** [review]
looks good

**Pull request merged**